MVP Version 6

nitaliano opened the pull request:
Just creating a PR to allow making code reviews easier, I don't plan on merging this in. Would love to get some feedback on the code so far.

**cc** @bsudekum @1ec5 @boundsj @zugaldia @tobrun

Commit:
Removed old android and ios projects

Commit:
Removed old android and ios projects

Commit:
Added start of javascript components

Commit:
Added start of android rewrite

Commit:
Updated gitignore and package.json

Commit:
Start of iOS project rewrite

Commit:
Added start of autogenerating docs

Commit:
Added eslint

Commit:
Some file cleanup

Commit:
Updated circle to lint

Commit:
Android file cleanup

Commit:
Added support for onPress and onLongPress

Commit:
Updated docs

Commit:
Added support for min/max zoom level, scroll/pitch enabled disabled

Commit:
Added support for map change events and flyTo

Commit:
Added complete callback for flyTo

Commit:
Map events and methods

Commit:
Updated docs

Commit:
Added example app to repo

Commit:
Version bump to 6.0.0

Commit:
Version bump to 6.0.0

Commit:
Updated build badge to point to v6 on circle for the time being

Commit:
Removed old projects after rebase

bsudekum commented on example/src/App.js:
```diff
@@ -0,0 +1,172 @@
+    if (IS_ANDROID) {
+      await MapboxGL.requestPermissions();
+    }
+    MapboxGL.setAccessToken('pk.eyJ1Ijoibmlja2l0YWxpYW5vIiwiYSI6ImNqNjVtemx3dzAyeHkyd3IwOHJ3b3JsNmMifQ.QSvBB90og-7PW68dZPDAIg');
```
Don't commit access tokens.

nitaliano replied:
Good catch! Updated

Commit:
Updated example to not have hardcoded token

bsudekum commented on example/src/App.js:
```diff
@@ -0,0 +1,173 @@
+    if (IS_ANDROID) {
+      await MapboxGL.requestPermissions();
+    }
+    MapboxGL.setAccessToken(config.get('accessToken'));
```
I don't think we should be dictating to developers how they should be configuring private/config variables. I think inserting the string `INSERT MAPBOX ACCESS TOKEN HERE. Your access token can be found here: Link` should be enough.

nitaliano replied:
My only concern having it as a hardcoded string would be that every time someone wants to add an example they would have to add the token here and remove the token here before making a commit.

bsudekum commented on example/src/components/SetUserTrackingModes.js:
```diff
@@ -0,0 +1,49 @@
+    this._trackingOptions = Object.keys(MapboxGL.UserTrackingModes).map((key) => {
+      return {
+        label: key,
+        data: MapboxGL.UserTrackingModes[key],
```
`trackingMode` might be more descriptive here.

bsudekum commented on ios/RCTMGL/MGLModule.m:
```diff
@@ -0,0 +1,78 @@
+    [styleURLS setObject:[MGLStyle.darkStyleURL absoluteString] forKey:@"Dark"];
+    [styleURLS setObject:[MGLStyle.lightStyleURL absoluteString] forKey:@"Light"];
+    [styleURLS setObject:[MGLStyle.outdoorsStyleURL absoluteString] forKey:@"Outdoors"];
+    [styleURLS setObject:[MGLStyle.satelliteStyleURL absoluteString] forKey:@"Satellite"];
```
There are a few more styles missing from this list: https://www.mapbox.com/api-documentation/#styles

nitaliano replied:
Updated

bsudekum commented on ios/RCTMGL/MGLModule.m:
```diff
@@ -0,0 +1,78 @@
+    [userTrackingModes setObject:[NSNumber numberWithInt:MGLUserTrackingModeNone] forKey:@"None"];
+    [userTrackingModes setObject:[NSNumber numberWithInt:MGLUserTrackingModeFollow] forKey:@"Tracking"];
+    [userTrackingModes setObject:[NSNumber numberWithInt:MGLUserTrackingModeFollowWithHeading] forKey:@"Compass"];
+    [userTrackingModes setObject:[NSNumber numberWithInt:MGLUserTrackingModeFollowWithCourse] forKey:@"Navigation"];
```
I might follow the tracking mode names more closely here. `.none`, `.follow`, `.followWithHeading`, `.followWithCourse`.

nitaliano replied:
This is where things will get interesting, the iOS and Android SDKs have different names for the modes. Unless I am looking in the wrong place.

https://github.com/mapbox/mapbox-plugins-android/blob/master/plugins/locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerMode.java

atarsha replied:
I ran into similar issues while working on a PR you recently merged. Android and iOS have different OfflinePackStates, so I had to create a method to normalize them into one. In that particular case, iOS had more states than Android, so it made sense to map them to the iOS states. In this particular case, I agree with @bsudekum. The iOS names are more descriptive. The actual result here isn't "compass" it's "follow the user with compass rotation." I also find "Tracking" vs "Compass" not very self-descriptive. While I know the answer to these questions, here is an example: If following the user is "tracking" then is "compass" not following the user and just rotating?

Ideally, this would be fixed upstream, but since it isn't, I think we should run with more descriptive, even if it means choosing our own names entirely for RN.

nitaliano replied:
I'll switch the names over. I agree we should be using more descriptive names.

bsudekum commented on ios/RCTMGL/RCTMGLMapViewManager.m:
```diff
@@ -0,0 +1,287 @@
+
+// prevents SDK from crashing and cluttering logs
+// since we don't have access to the frame right away
+static CGRect const RCT_MAPBOX_MIN_MAP_FRAME = { { 0.0f, 0.0f }, { 64.0f, 64.0f } };
```
Would it be possible to make this equal to the height/width of the parent view?

nitaliano replied:
We don't have access to that information unfortunately when we initialize the view.  It hasn't been inserted into the view hierarchy at this point. RN behind the scenes on iOS will mess around with this view and set the frame based off it's parent and `style` attribute passed down from the JS later on once it has access to the information.

jackfengji commented on android/rctmgl/src/main/AndroidManifest.xml:
```diff
@@ -1,7 +1,5 @@
-<?xml version="1.0" encoding="utf-8"?>
-<manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    package="com.mapbox.reactnativemapboxgl" >
+<manifest xmlns:android="http://schemas.android.com/apk/res/android" package="mapbox.rctmgl">
```
Why do you use mapbox.rctmgl instead of com.mapbox.rctmgl?

nitaliano replied:
It should be `com.mapbox.rctmgl` good catch

nitaliano replied:
Just pushed a commit to fix the package

jackfengji commented on android/rctmgl/src/main/java/mapbox/rctmgl/components/mapview/RCTMGLMapViewManager.java:
```diff
@@ -0,0 +1,180 @@
+                .put(EventKeys.MAP_CLICK, "onPress")
+                .put(EventKeys.MAP_LONG_CLICK,"onLongPress")
+                .put(EventKeys.MAP_ONCHANGE, "onMapChange")
+                .put(EventKeys.USER_LOCATION_CHANGE, "onUserLocationChange")
```
I don't think we should have this event in mapbox SDK. User should listen to location changes from native SDKs.

nitaliano replied:
I'm all for dropping this the less updates going over the bridge from this repo the better.

atarsha replied:
Just a reminder forget to remove that event from docs/MapView.md as well.

EDIT: Disregard, I see that is a result of auto-generating docs, and will be removed automatically once removed from the code.

jackfengji commented on android/rctmgl/src/main/java/mapbox/rctmgl/components/mapview/RCTMGLMapView.java:
```diff
@@ -0,0 +1,438 @@
+        if (bounds == null) {
+            return;
+        }
+        mMap.easeCamera(CameraUpdateFactory.newLatLngBounds(bounds, padding), durationMS);
```
Do we need to handle the situation that user calls fitBounds/flyTo before map ready? For example in componentDidMount

nitaliano replied:
Yes, we absolutely should.

nitaliano replied:
Updated the PR, I put all camera updates on a queue now so something like this is now possible

```
this.map.setCamera({
  stops: [
    { pitch: 45, duration: 200 },
    { centerCoordinates: [...], zoom: 16, mode: MapboxGL.CameraModes.Flight },
    { heading: 180, duration: 2000 },
    { heading: 360, duration: 2000 },
    {
       bounds: {
          ne: [...],
          sw: [...],
          padding: 15
       }
    }
  ]
})
```

Commit:
Updated android package name and made RCTMGLMapView inherit MapView

Commit:
Added queue for camera updates and ability to set camera stops

Commit:
Added missing base style urls

Commit:
Updated user tracking enum names

Commit:
Updated eslint rules to ignore child props

Commit:
Updated script to also lint example project

Commit:
Linted example project

Commit:
Added support for zoomTo

1ec5 commented on ios/RCTMGL/RCTMGLMapView.m:
```diff
@@ -78,6 +93,30 @@ - (void)setReactUserTrackingMode:(int)reactUserTrackingMode
+                                                 max:[NSNumber numberWithDouble:LAT_MAX]] doubleValue];
+    
+    double constrainedScale = pow(2.0, constrainedZoom);
+    return cos(constrainedLatitude * DEG2RAD) * M2PI * EARTH_RADIUS_M / (constrainedScale * TILE_SIZE);
```
I suspect you could take the result of `-[MGLMapView metersPerPointAtLatitude:]` and scale it by the zoom level without having to perform the above math independently.

1ec5 commented on ios/RCTMGL/RCTMGLMapView.m:
```diff
@@ -78,6 +93,30 @@ - (void)setReactUserTrackingMode:(int)reactUserTrackingMode
 
+#pragma mark - methods
+
+- (CLLocationDistance)getMetersPerPixelAtLatitude:(double)latitude withZoom:(double)zoomLevel
```
The usual case would be to convert from pixels to meters at the current zoom level. Maybe make the zoom level optional?

1ec5 commented on ios/RCTMGL/RCTMGLMapView.m:
```diff
@@ -78,6 +93,30 @@ - (void)setReactUserTrackingMode:(int)reactUserTrackingMode
+    CLLocationDistance metersPerPixel = [self getMetersPerPixelAtLatitude:self.camera.centerCoordinate.latitude withZoom:zoomLevel];
+    CLLocationDistance metersTall = metersPerPixel * self.frame.size.height;
+    CLLocationDistance altitude = metersTall / 2 / tan(MGLRadiansFromDegrees(30) / 2.0);
+    return altitude * sin(M_PI_2 - MGLRadiansFromDegrees(self.camera.pitch)) / sin(M_PI_2);
```
mapbox/mapbox-gl-native#5583 would be nice to have.

nitaliano replied:
Yes it would!

Commit:
Removed user location change events

Commit:
Added support to autogen react method markdown

Commit:
Updated docs

Commit:
Added v8 style spec

Commit:
Script to autogenerate styles and docs

Commit:
Added base layer and source components

Commit:
Added autogenerated style code

Commit:
Added android style layers and sources

Commit:
Added ios style layers and sources

Commit:
Added Mapbox StyleSheet

Commit:
Added javascript style layers and components

Commit:
Added style layer and source examples

Commit:
Added docs

Commit:
Added start of JS unit tests

Commit:
Updated dot files and package.json

Commit:
Removed old unused android file

Commit:
Left over after merge

Commit:
Fix unit tests

Commit:
Lint

Commit:
Update circle to work with yarn

Commit:
retrigger build changed to use ubunut 14 instead of 12

Commit:
Add eslint override to example project

Commit:
Added support for background layer

Commit:
Added support for RasterLayer and RasterSource

Commit:
Added support for Light

Commit:
Updated autogen script and code to support Light

Commit:
Added support for styles to be updated dynamically

Commit:
Updated docs

Commit:
Added more examples

Commit:
Lint and updated stylesheet unit tests

Commit:
Added all style enums as constants on module

Commit:
Updated example project to work with npm scope change

Commit:
Update documentation

Commit:
Add alpha notice

Commit:
Update app.iml

Commit:
Updated example install documentation

Commit:
Added extra steps to android install

Commit:
Bumped to alpha1

nitaliano committed:
Fix typo in readme

Commit:
Updated HEADER_SEARCH_PATHS to find react now that we are under @mapbox

Commit:
Updated RCTMGLStyle to init with an MGLStyle to make setting images easier

Commit:
Added support for attributionEnabled, logoEnabled, and rotateEnabled on mapview

Commit:
Monkey patched android callback support

Commit:
Updated flyTo example to use a promise on completion

Commit:
Added an icon symbol layer example

Commit:
Added support for feature querying API

jliebrand committed:
Minor tweaks

Tripped me up, so thought I'd update the readme...

Commit:
Updated fitBounds to also accept an array so we can control padding on each side

Commit:
Update RCTLayer -> RCTMGLLayer and RCTSource -> RCTMGLLayer to prevent naming collisions with RN

Commit:
Bumped version to 6.0.0.alpha3

Commit:
Updated example to use RN 0.49

Commit:
Lint

Commit:
Add iOS install step about linking libRCTMGL.a

Commit:
Use PermissionAndroid from RN instead of using Mapbox SDK permission manager

Commit:
Updated interface teat and example app to use new permission method

Commit:
Normalize what None means on android for user location

kevinsperrine committed:
Better version tracking on Mapbox.framework downloads

Added a `.framework_version` file to check against and only delete and re-download the Mapbox.framework when the version has changed.

kevinsperrine committed:
update .gitignore to exclude .framework_version

kevinsperrine committed:
Added "Downloading" message back if new download is required

Commit:
Added android support for iconImage token

Commit:
Added ios support for iconImage token

Commit:
Update ShapeSouce JS component to include images

Commit:
Added exampe

Commit:
Updated babel-preset-react-native version and fix indentations in package.json

Commit:
Cleanup images once we remove source on Android

Commit:
Add example app.iml to .gitignore

Commit:
Fixes SymbolLayer textOffset and iconOffset on iOS

There seem to be some styles that will be special cases to the rules used from the json spec
this PR adds the ability to add overrides, so we can be flexible in special cases. In this case
we were using a NSArray for textOffset and iconOffset when it should be a CGVector.

Commit:
Adds support for PointAnnotations and Callouts on Android

This adds the ability to display custom React views in a PointAnnotation and Callout

Commit:
Adds support for PointAnnotations and Callouts on iOS

This adds the ability to render React Views in PointAnnotations and Callouts

Commit:
Adds PointAnnotation and Callout JS components

Commit:
Adds PointAnnotation example using custom views

Commit:
Added example resources

Commit:
Adds support for native assets as an iconImage

If you want to use any images in your geojson that are under your
drawables dir on Android and *.xcassets on iOS you now can provide
an assets key to the images prop for the ShapeSource and the native code
will add these images to the map style for you.

frankrowe committed:
change eventtype

Commit:
Added support for sourceLayerID on FillLayer

Adds the missing sourceLayerID property to our FillLayer component and
added an example showing how to use a custom vector source with a FillLayer.

Commit:
Added global image queue for downloading images on iOS

We added a image queue backed by NSOperationQueue which handles threading for us
and now all image downloads will be places on the queue, instead of using a dispatch group
or semaphore

Commit:
Added iconAllowOverlap to custom icon example

Commit:
Updated the zoomTo call to use a promise instead of callback

Commit:
Updated Show Map example to follow user location

Commit:
Made Callout methods private

Commit:
Added docs for PointAnnotation and Callout

Commit:
Added PointAnnotation and Callout links to README

Commit:
Update license year

Commit:
Updated package.json and example package json version

nitaliano committed:
Merge pull request #686 from mapbox/v6-beta

[V6] Beta

Commit:
Added resolveAssetSource mock from RN

Commit:
Added image asset resolving unit test for StyleSheet

This unit test will test if the image was required(import, require) by the JS layer
that it will resolve it to an asset/network uri

Commit:
Added helper function to StyleSheet to decide when to resolve images

Commit:
Updates how we add images to map style

We were missing the use case when a user wants to use a custom sprite
this now checks to make sure that we only add an image to the map style
if it has a valid uri/url

Commit:
Remove isTokenString check now that we validate URIs

nitaliano committed:
Merge pull request #689 from mapbox/icon-sprite-sheet-fix

[v6] Icon sprite sheet fix

Commit:
Adds support for offline packs in Android

Commit:
Adds support for offline packs in iOS

Commit:
Adds javascript interface for offline packs

Commit:
offlineManager unit tests

Commit:
Updated autogen docs to handle modules

Commit:
Added offlineManager documentation

Commit:
Added example to create/pause/resume an offline region downloads

Commit:
Update Mapbox.framework version

Commit:
Added support for compassEnabled and contentInset

Commit:
Adds support for visible region in onRegion change events

Now a property called visibleBounds will be included with all region change events

Commit:
Added method getVisibleBounds on the MapView component

Commit:
Update FillLayer to override addToMap to work with source layers

Commit:
Update Android SourceManagers to manage children

VectourSource and ShapeSource have been updated to manage their own
children in a similar way as the MapView to have finer control over
when to remove and add a layer to MapboxMap

Commit:
Updated cloneReactChildrenWithProps to be able to handle null children

Commit:
Updated the add to map layer flow

Commit:
Updated cloneReactChildrenWithProps to handle single child

nitaliano committed:
Merge pull request #696 from mapbox/feature/v6-offline

[v6] Offline Support

Commit:
Fix typo in Android autogen script

Commit:
Use existing layers/sources

Instead of just checking to see if a layer or source exists in the base composite
update to just look up by id even if it's not the composite to help with more
custom use cases

Commit:
Made StyleSheet, Layers, and Sources compatible with Animated

Commit:
Unlinked deps that are not needed

Commit:
Added example of animating along a line

Commit:
Fixes some iOS warnings

Commit:
Fixes JS warnings

This fixes JS warnings now that we are using requireNativeComponent
correctly.

Commit:
Fix android warnings

Commit:
Update animation example to show progress line

Commit:
Add ConstantPropType and added vector override for textTranslate

Commit:
Android - Invalidate visible markers in region onMapReady

Commit:
Android - added lifecycle events to dispose, pause, resume GPS

Commit:
Bump to beta2

nitaliano committed:
Merge pull request #705 from mapbox/v6-improvements

[WIP] [V6] Improvements
